## Toronto Neighborhood Analysis

Importing all the required libraries!

In [1]:
import requests
import lxml.html as lh
import pandas as pd
from pandas.io.json import json_normalize
from sklearn.cluster import KMeans
!pip install folium
import folium
from IPython.display import display

Using the requests library to get the HTML content from the wikipedia webpage and then extracting all the data which are in the **tr** block of the HTML as those are the rows of the table. 

In [2]:
URL = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
toronto_page = requests.get(URL)

doc = lh.fromstring(toronto_page.content)

rows = doc.xpath('//tr')

The first row in the extracted rows is the name of columns. So extracting it and storing it in a list. 

In [3]:
cols = []

for column in rows[0]:
    cols.append(column.text_content()[:-1])
    
cols

['Postal Code', 'Borough', 'Neighborhood']

Getting all other rows and storing it in the list variable named _'data'_.

In [4]:
data = []

for index, r in enumerate(rows):
    if index > 0 and index < 181:    # Index 180 marks the end of the table. 
        row = []
        for value in r:
            row.append(value.text_content()[:-1])
        data.append(row)

The **data** list is converted to a pandas dataframe and every _Borough_ which is _Not assigned_ is removed from the dataframe. 

In [5]:
toronto_neighborhood = pd.DataFrame(data, columns=cols)
toronto_neighborhood = toronto_neighborhood[toronto_neighborhood['Borough'] != 'Not assigned'].reset_index(drop=True)

toronto_neighborhood.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [6]:
toronto_neighborhood.shape

(103, 3)

In [7]:
print(toronto_neighborhood['Postal Code'].nunique())
print("-----------------")
print(toronto_neighborhood[toronto_neighborhood['Neighborhood'] == 'Not assigned'].count())

103
-----------------
Postal Code     0
Borough         0
Neighborhood    0
dtype: int64


As there are total 103 rows in the table and also 103 unique values of postal codes, so there are no two postal codes which are the same. Also, there are no neighborhoods with **Not assigned** entry. 

Downloading the geodata of toronto neighborhoods!

In [8]:
!wget http://cocl.us/Geospatial_data

--2020-07-02 06:58:05--  http://cocl.us/Geospatial_data
Resolving cocl.us (cocl.us)... 158.85.108.83, 158.85.108.86, 169.48.113.194
Connecting to cocl.us (cocl.us)|158.85.108.83|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://cocl.us/Geospatial_data [following]
--2020-07-02 06:58:06--  https://cocl.us/Geospatial_data
Connecting to cocl.us (cocl.us)|158.85.108.83|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-07-02 06:58:08--  https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.box.com (ibm.box.com)... 107.152.29.197
Connecting to ibm.box.com (ibm.box.com)|107.152.29.197|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /public/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-07-02 06:58:09--  https://ibm.box.com/public

Merging the two dataframes. 

In [9]:
geodata = pd.read_csv("Geospatial_data")
toronto_geodata = pd.merge(toronto_neighborhood, geodata, how='inner')

toronto_geodata.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Analysing all the boroughs wich have the word **York** in it. 

In [10]:
york_data = toronto_geodata[toronto_geodata['Borough'].apply(lambda x: 'York' in x)].reset_index(drop = True)

## Exploring Places near Neighborhoods

Using Foursquare API Credentials

In [11]:
CLIENT_ID = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXX' 
CLIENT_SECRET = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX' 
VERSION = '20180604'

Limiting the results of the API to a 100 per venue, and exploring around the radius of 1000 meters per venue. 

In [12]:
LIMIT = 100
radius = 1000
venues = []

for latitude, longitude in zip(york_data['Latitude'], york_data['Longitude']):
    url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
    venues.append(requests.get(url).json())

In [13]:
york_popular = pd.DataFrame(columns=['Postal Code', 'Borough', 'Neighborhood', 'Latitude', 'Longitude', 'Place'])
row_nums = 0

for index, row in enumerate(venues):
    try:
        for i in range(len(row['response']['groups'][0]['items'])):
            york_popular.loc[row_nums] = list(york_data.iloc[index]) + [row['response']['groups'][0]['items'][i]['venue']['categories'][0]['name']]
            row_nums += 1
    except:
        pass
    
york_popular.head()

Postal Code     Borough Neighborhood   Latitude  Longitude  \
0         M3A  North York    Parkwoods  43.753259 -79.329656   
1         M3A  North York    Parkwoods  43.753259 -79.329656   
2         M3A  North York    Parkwoods  43.753259 -79.329656   
3         M3A  North York    Parkwoods  43.753259 -79.329656   
4         M3A  North York    Parkwoods  43.753259 -79.329656   

                  Place  
0  Caribbean Restaurant  
1                  Park  
2                  Café  
3         Grocery Store  
4     Fish & Chips Shop

Now we have the places around each neighborhood with the help of the Foursquare API. After that I try to find the eight most common venues for each neighborhood.

In [14]:
york_onehot = pd.get_dummies(york_popular['Place'])
york_onehot.insert(0, 'Neighborhood', york_popular['Neighborhood'])

york_grouped = york_onehot.groupby('Neighborhood').mean().reset_index()

york_sorted = pd.DataFrame(columns=['Neighborhood', '1', '2', '3', '4', '5', '6', '7', '8'])
york_sorted['Neighborhood'] = york_grouped['Neighborhood']

for index in range(york_grouped.shape[0]):
    temp = york_grouped.iloc[index, 1:]
    temp_sort = temp.sort_values(ascending=False)
    york_sorted.iloc[index, 1:] = list(temp_sort.index.values[0:8])

york_sorted.head()

Neighborhood                       1  \
0   Bathurst Manor, Wilson Heights, Downsview North             Pizza Place   
1                                   Bayview Village           Grocery Store   
2                 Bedford Park, Lawrence Manor East             Coffee Shop   
3                               Caledonia-Fairbanks                Pharmacy   
4  Del Ray, Mount Dennis, Keelsdale and Silverthorn  Furniture / Home Store   

                    2                    3                     4  \
0         Coffee Shop                 Bank            Ski Chalet   
1                Bank  Japanese Restaurant           Gas Station   
2  Italian Restaurant           Restaurant        Sandwich Place   
3                Park             Bus Stop   Japanese Restaurant   
4       Grocery Store       Discount Store  Fast Food Restaurant   

                           5                  6                         7  \
0  Middle Eastern Restaurant      Shopping Mall  Mediterranean Restaurant   
1                       Park              Trail        Chinese Restaurant   
2                       Bank  Indian Restaurant                   Butcher   
3              Grocery Store     Cosmetics Shop               Coffee Shop   
4                Coffee Shop         Restaurant              Dessert Shop   

                8  
0  Sandwich Place  
1         Dog Run  
2       Pet Store  
3  Discount Store  
4  Sandwich Place

Based on the frequency of venues, I divide them into four different cluster using the KMeans Clustering Method. 

In [15]:
model = KMeans(init='k-means++', n_clusters=4, random_state=0).fit(york_grouped.iloc[:, 1:])

york_sorted.insert(1, 'Cluster Label', model.labels_)

york_final = york_data.copy()
york_final = york_final.join(york_sorted.set_index('Neighborhood'), on='Neighborhood')
york_final.dropna(axis=0, inplace=True)

york_final.head()

Postal Code     Borough                      Neighborhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M4B   East York   Parkview Hill, Woodbine Gardens  43.706397   

   Longitude  Cluster Label                    1                       2  \
0 -79.329656            1.0                 Park                Pharmacy   
1 -79.315572            3.0          Coffee Shop   Portuguese Restaurant   
2 -79.464763            3.0       Clothing Store  Furniture / Home Store   
3 -79.352188            3.0  Japanese Restaurant             Coffee Shop   
4 -79.309937            3.0          Pizza Place    Fast Food Restaurant   

                      3                  4                 5  \
0         Shopping Mall  Convenience Store          Bus Stop   
1   Sporting Goods Shop       Hockey Arena       Golf Course   
2  Fast Food Restaurant        Coffee Shop  Sushi Restaurant   
3           Pizza Place       Burger Joint              Bank   
4  Gym / Fitness Center            Brewery      Intersection   

                       6                    7               8  
0                   Road      Laundry Service  Shop & Service  
1           Intersection    French Restaurant     Men's Store  
2  Vietnamese Restaurant  Fried Chicken Joint      Restaurant  
3            Supermarket      Thai Restaurant  Breakfast Spot  
4     Rock Climbing Spot       Breakfast Spot  Soccer Stadium

## Visualization

All the clusters made are visualized on the map below using the Folium library.  

In [16]:
latitude = 43.6957
longitude = -79.4504

york_map = folium.Map(location=[latitude, longitude], zoom_start=12)

colors = ['red', 'blue', 'green', 'yellow']

for lat, lon, neighborhood, cluster in zip(york_final['Latitude'], york_final['Longitude'], york_final['Neighborhood'], york_final['Cluster Label']):
    label = folium.Popup(str(neighborhood))
    folium.CircleMarker(
        [lat, lon], 
        radius=5,
        popup=label,
        color=colors[int(cluster)],
        fill=True, 
        fill_opacity=0.7
    ).add_to(york_map)
    
display(york_map)

## Examiming the clusters

In [17]:
york_final[york_final['Cluster Label'] == 0]

Postal Code     Borough      Neighborhood   Latitude  Longitude  \
27         M9M  North York  Humberlea, Emery  43.724766 -79.532242   

    Cluster Label                  1             2                 3  \
27            0.0  Convenience Store  Intersection  Storage Facility   

                 4     5            6       7            8  
27  Discount Store  Park  Gas Station  Bakery  Golf Course

We can see that in the next cluster, all the neighborhoods have Park and Pharmacies among their most common venues. While some of them share Pizza Places, Bus Stops, Coffee Shops, etc. as their most common venues.  

Similarly, we can see common characteristics within other clusters. 

In [18]:
york_final[york_final['Cluster Label'] == 1]

Postal Code     Borough                 Neighborhood   Latitude  Longitude  \
0          M3A  North York                    Parkwoods  43.753259 -79.329656   
9          M6E        York          Caledonia-Fairbanks  43.689026 -79.453512   
11         M2H  North York            Hillcrest Village  43.803762 -79.363452   
22         M9L  North York                Humber Summit  43.756303 -79.565963   
33         M2R  North York  Willowdale, Willowdale West  43.782736 -79.442259   

    Cluster Label         1            2              3                    4  \
0             1.0      Park     Pharmacy  Shopping Mall    Convenience Store   
9             1.0  Pharmacy         Park       Bus Stop  Japanese Restaurant   
11            1.0  Pharmacy  Coffee Shop           Park       Ice Cream Shop   
22            1.0    Bakery  Pizza Place       Pharmacy  Arts & Crafts Store   
33            1.0  Pharmacy       Bakery     Baby Store       Discount Store   

                                           5               6                7  \
0                                   Bus Stop            Road  Laundry Service   
9                              Grocery Store  Cosmetics Shop      Coffee Shop   
11  Residential Building (Apartment / Condo)            Bank           Bakery   
22                         Electronics Store   Shopping Mall             Park   
33               Eastern European Restaurant            Bank             Park   

                     8  
0       Shop & Service  
9       Discount Store  
11   Recreation Center  
22  Italian Restaurant  
33         Coffee Shop

In [19]:
york_final[york_final['Cluster Label'] == 2]

Postal Code     Borough              Neighborhood  Latitude  Longitude  \
19         M2L  North York  York Mills, Silver Hills  43.75749 -79.374714   

    Cluster Label     1     2            3        4               5  \
19            2.0  Park  Pool  Yoga Studio  Dog Run  Farmers Market   

                     6            7                     8  
19  Falafel Restaurant  Event Space  Ethiopian Restaurant

In [20]:
york_final[york_final['Cluster Label'] == 3]

Postal Code     Borough                                      Neighborhood  \
1          M4A  North York                                  Victoria Village   
2          M6A  North York                  Lawrence Manor, Lawrence Heights   
3          M3B  North York                                         Don Mills   
4          M4B   East York                   Parkview Hill, Woodbine Gardens   
5          M6B  North York                                         Glencairn   
6          M3C  North York                                         Don Mills   
7          M4C   East York                                  Woodbine Heights   
8          M6C        York                                Humewood-Cedarvale   
10         M4G   East York                                           Leaside   
12         M3H  North York   Bathurst Manor, Wilson Heights, Downsview North   
13         M4H   East York                                  Thorncliffe Park   
14         M2J  North York                      Fairview, Henry Farm, Oriole   
15         M3J  North York                   Northwood Park, York University   
16         M4J   East York     East Toronto, Broadview North (Old East York)   
17         M2K  North York                                   Bayview Village   
18         M3K  North York                                         Downsview   
20         M3L  North York                                         Downsview   
21         M6L  North York          North Park, Maple Leaf Park, Upwood Park   
23         M2M  North York                           Willowdale, Newtonbrook   
24         M3M  North York                                         Downsview   
25         M5M  North York                 Bedford Park, Lawrence Manor East   
26         M6M        York  Del Ray, Mount Dennis, Keelsdale and Silverthorn   
28         M2N  North York                       Willowdale, Willowdale East   
29         M3N  North York                                         Downsview   
31         M9N        York                                            Weston   
32         M2P  North York                                   York Mills West   

     Latitude  Longitude  Cluster Label                       1  \
1   43.725882 -79.315572            3.0             Coffee Shop   
2   43.718518 -79.464763            3.0          Clothing Store   
3   43.745906 -79.352188            3.0     Japanese Restaurant   
4   43.706397 -79.309937            3.0             Pizza Place   
5   43.709577 -79.445073            3.0           Grocery Store   
6   43.725900 -79.340923            3.0     Japanese Restaurant   
7   43.695344 -79.318389            3.0                    Park   
8   43.693781 -79.428191            3.0             Pizza Place   
10  43.709060 -79.363452            3.0             Coffee Shop   
12  43.754328 -79.442259            3.0             Pizza Place   
13  43.705369 -79.349372            3.0             Coffee Shop   
14  43.778517 -79.346556            3.0          Clothing Store   
15  43.767980 -79.487262            3.0             Coffee Shop   
16  43.685347 -79.338106            3.0             Coffee Shop   
17  43.786947 -79.385975            3.0           Grocery Store   
18  43.737473 -79.464763            3.0   Vietnamese Restaurant   
20  43.739015 -79.506944            3.0   Vietnamese Restaurant   
21  43.713756 -79.490074            3.0             Coffee Shop   
23  43.789053 -79.408493            3.0       Korean Restaurant   
24  43.728496 -79.495697            3.0   Vietnamese Restaurant   
25  43.733283 -79.419750            3.0             Coffee Shop   
26  43.691116 -79.476013            3.0  Furniture / Home Store   
28  43.770120 -79.408493            3.0             Coffee Shop   
29  43.761631 -79.520999            3.0   Vietnamese Restaurant   
31  43.706876 -79.518188            3.0             Coffee Shop   
32  43.752758 -79.400049            3.0             Coffee Shop   

                         2      